DWT-1075 Add service-to-service authentication plugin

abeyonalaja opened the pull request:
### Description
Ticket [DWT-1075](https://eaflood.atlassian.net/browse/DWT-1075)

- Introduced a `service-auth` plugin to enable bearer token-based authentication for protected routes.
- Updated applicable routes to integrate the new authentication mechanism.
- Added tests to validate the service authentication functionality.
- Implement

[DWT-1075]: https://eaflood.atlassian.net/browse/DWT-1075?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

abeyonalaja committed:
Add service-to-service authentication plugin

Introduced a `service-auth` plugin to enforce bearer token-based authentication for protected routes. Updated relevant routes and tests to integrate and validate the authentication mechanism. Added support for a configurable service authentication token.

shepda commented on src/plugins/service-auth.js:
```diff
@@ -0,0 +1,40 @@
+      server.auth.scheme('service-token-scheme', () => {
+        return {
+          authenticate: (request, h) => {
+            const authorization = request.headers.authorization
```
In the API the casing of the authorization header is different to what is being checked here `Authorization` v `authorization`.

Would that cause this check to fail?

The unit tests don't test this difference in casing as the header test data has this property in lowercase, which is the same as is used in this check.

shepda commented on src/plugins/service-auth.js:
```diff
@@ -0,0 +1,40 @@
+
+            const [scheme, token] = authorization.split(' ')
+
+            if (scheme?.toLowerCase() !== 'bearer' || !token) {
```
Appreciate that this may be left open to allow for different string formats but as UUID is specified in the `serviceAuthToken` config property in the API, would it be worth checking that the `token` value is a valid UUID here as an initial check?

shepda replied:
A UUID isn't an actual bearer token but the standard format for a bearer auth field is `Bearer xxxxx` so I don't think it would be unreasonable to remove the scheme conversion to lowercase and expect the casing to be exact.

abeyonalaja committed:
Replace custom auth scheme with @hapi/basic

Replaced the custom service-token authentication scheme with @hapi/basic for improved maintainability and adherence to standards. Updated tests and routes to use Basic Authentication headers instead of Bearer tokens, and added the @hapi/basic package as a dependency.

abeyonalaja committed:
Refactor auth to use clientId instead of serviceAuthToken

Replaced serviceAuthToken configuration with clientId for authentication. Updated validation logic and tests to reflect this change, ensuring clientId is validated and propagated in credentials. Removed unused serviceAuthToken from configuration and related imports.

abeyonalaja committed:
Add service-to-service authentication with credential validation

Introduced a new system to validate service-to-service authentication using basic credentials. Added custom validation logic, updated configuration to handle credentials, and enhanced plugin behavior for local and non-local environments with corresponding tests.

abeyonalaja committed:
Allow null service credentials for non-local environments

Updated service authentication logic to handle null credentials gracefully in non-local environments. Adjusted validation and coercion in `validate-service-credentials` to support nullable values. Added tests and updated configurations to reflect new behavior.

abeyonalaja committed:
Refactor conditional statements in validation logic

Wrapped conditional statements in braces for better readability and consistency. This change improves the code style and aligns with standard best practices for maintainability.

abeyonalaja committed:
Remove local environment auth bypass and update test cases

The auth plugin now consistently validates credentials regardless of environment. Tests were updated to reflect this change, removing references to the local environment bypass and ensuring proper validation logic for all scenarios.

abeyonalaja committed:
Add auth-related environment variables to services

Introduced `SERVICE_AUTH_PASSWORD` and `ACCESS_CRED_WASTE_MOVEMENT_EXTERNAL_API` environment variables in `compose.yml` and workflow files. These changes enable authentication configurations for the services and CI workflows, improving security and functionality.

abeyonalaja committed:
Use environment variables for sensitive credentials

Replaced hardcoded sensitive values in `compose.yml` with environment variables for improved security and flexibility. Updated `.env.example` to include placeholders for the new variables.

abeyonalaja committed:
Remove service-auth plugin and integrate authentication in server.js

The service-auth plugin was removed, and its functionality was directly implemented in server.js. This simplifies the codebase by reducing indirection and centralizing the authentication logic. Corresponding tests were also deleted as they became redundant.
